File: src/test/java/edu/hm/hafner/analysis/parser/MavenParserTest.java
package edu.hm.hafner.analysis.parser;

import java.util.Iterator;

import edu.hm.hafner.analysis.AbstractParser;
import edu.hm.hafner.analysis.AbstractParserTest;
import edu.hm.hafner.analysis.Issue;
import edu.hm.hafner.analysis.Issues;
import edu.hm.hafner.analysis.Priority;
import edu.hm.hafner.analysis.assertj.IssueAssert;
import edu.hm.hafner.analysis.assertj.SoftAssertions;
import static edu.hm.hafner.analysis.assertj.SoftAssertions.*;

/**
 * Tests the class {@link JavacParser} for output log of a maven compile.
 */
public class MavenParserTest extends AbstractParserTest {

    /**
     * Creates a new instance of {@link MavenParserTest}.
     *
     */
    protected MavenParserTest() {
        super("maven.txt");
    }

    protected String getWarningsFile() {
        return "maven.txt";
    }

    @Override
    protected void assertThatIssuesArePresent(final Issues<Issue> issues, final SoftAssertions softly) {
        softly.assertThat(issues)
                .hasSize(5);

        Iterator<Issue> iterator = issues.iterator();

        assertThatWarningIsAtLine(softly, iterator.next(), 3);
        assertThatWarningIsAtLine(softly, iterator.next(), 36);
        assertThatWarningIsAtLine(softly, iterator.next(), 47);
        assertThatWarningIsAtLine(softly, iterator.next(), 69);
        assertThatWarningIsAtLine(softly, iterator.next(), 105);
    }

    private IssueAssert assertThatWarningIsAtLine(final SoftAssertions softly, final Issue warning, final int lineNumber) {
        return softly.assertThat(warning)
                .hasPriority(Priority.NORMAL)
                .hasCategory(AbstractParser.PROPRIETARY_API)
                .hasLineStart(lineNumber)
                .hasLineEnd(lineNumber)
                .hasMessage("com.sun.org.apache.xerces.internal.impl.dv.util.Base64 is Sun proprietary API and may be removed in a future release")
                .hasFileName("/home/hudson/hudson/data/jobs/Hudson main/workspace/remoting/src/test/java/hudson/remoting/BinarySafeStreamTest.java");
    }

    @Override
    protected AbstractParser createParser() {
        return new JavacParser();
    }
}

File: src/test/java/edu/hm/hafner/analysis/parser/MavenParserTest.kt
package edu.hm.hafner.analysis.parser

import edu.hm.hafner.analysis.*
import edu.hm.hafner.analysis.assertj.assertSoftly
import edu.hm.hafner.analysis.assertj.assertThat
import java.io.IOException

import org.junit.jupiter.api.Test

import org.junit.jupiter.api.Assertions.*

/**
 * Tests the class [JavacParser] for output log of a maven compile.
 */
open class MavenParserTest : ParserTester() {

    /**
     * Parses a file with two deprecation warnings.
     *
     * @throws IOException if the file could not be read
     */
    @Test
    @Throws(IOException::class)
    fun parseMaven() {

        assertSoftly<IssuesSoftAssertions> {
            val warnings = JavacParser().parse(openFile())


            assertThat(warnings)
                    .hasSize(5)

            var index = 0

            assertThat(warnings[index++])
                    .hasLineStart(3)
                    .hasLineEnd(3)
                    .hasMessage("com.sun.org.apache.xerces.internal.impl.dv.util.Base64 is Sun proprietary API and may be removed in a future release")
                    .hasFileName("/home/hudson/hudson/data/jobs/Hudson main/workspace/remoting/src/test/java/hudson/remoting/BinarySafeStreamTest.java")
                    .hasType(WARNING_TYPE)
                    .hasCategory(AbstractParser.PROPRIETARY_API)
                    .hasPriority(Priority.NORMAL)

            assertThat(warnings[index++])
                    .hasLineStart(36)
                    .hasLineEnd(36)
                    .hasMessage("com.sun.org.apache.xerces.internal.impl.dv.util.Base64 is Sun proprietary API and may be removed in a future release")
                    .hasFileName("/home/hudson/hudson/data/jobs/Hudson main/workspace/remoting/src/test/java/hudson/remoting/BinarySafeStreamTest.java")
                    .hasType(WARNING_TYPE)
                    .hasCategory(AbstractParser.PROPRIETARY_API)
                    .hasPriority(Priority.NORMAL)

            assertThat(warnings[index++])
                    .hasLineStart(47)
                    .hasLineEnd(47)
                    .hasMessage("com.sun.org.apache.xerces.internal.impl.dv.util.Base64 is Sun proprietary API and may be removed in a future release")
                    .hasFileName("/home/hudson/hudson/data/jobs/Hudson main/workspace/remoting/src/test/java/hudson/remoting/BinarySafeStreamTest.java")
                    .hasType(WARNING_TYPE)
                    .hasCategory(AbstractParser.PROPRIETARY_API)
                    .hasPriority(Priority.NORMAL)


            assertThat(warnings[index++])
                    .hasLineStart(69)
                    .hasLineEnd(69)
                    .hasMessage("com.sun.org.apache.xerces.internal.impl.dv.util.Base64 is Sun proprietary API and may be removed in a future release")
                    .hasFileName("/home/hudson/hudson/data/jobs/Hudson main/workspace/remoting/src/test/java/hudson/remoting/BinarySafeStreamTest.java")
                    .hasType(WARNING_TYPE)
                    .hasCategory(AbstractParser.PROPRIETARY_API)
                    .hasPriority(Priority.NORMAL)


            assertThat(warnings[index])
                    .hasLineStart(105)
                    .hasLineEnd(105)
                    .hasMessage("com.sun.org.apache.xerces.internal.impl.dv.util.Base64 is Sun proprietary API and may be removed in a future release")
                    .hasFileName("/home/hudson/hudson/data/jobs/Hudson main/workspace/remoting/src/test/java/hudson/remoting/BinarySafeStreamTest.java")
                    .hasType(WARNING_TYPE)
                    .hasCategory(AbstractParser.PROPRIETARY_API)
                    .hasPriority(Priority.NORMAL)

        }
    }

    override fun getWarningsFile(): String {
        return "maven.txt"
    }

    companion object {
        private val WARNING_TYPE = JavacParser().id
    }
}

